### PR TITLE
Fixed TypedRangeStideSegment to work with strongly typed indices

### PR DESCRIPTION
**Pull request by ajkunen**
We didn't have any unit tests for this, so I've added some tests and fixed the code so it works.

### CODE REVIEW

File: include/RAJA/index/IndexValue.hpp [hunk 1]
```diff
@@ -145,12 +145,24 @@ struct IndexValue {
...
     return TYPE(value / a.value);
   }
 
+  //! modulus to underlying index from an Index_type
+  RAJA_HOST_DEVICE RAJA_INLINE TYPE operator%(Index_type a) const
+  {
+    return TYPE(value % a);
+  }
+
+  //! division to underlying index from another strong type
```
Comment (rhornung67) [inline include/RAJA/index/IndexValue.hpp]: This looks like a modulus to me...

File: include/RAJA/index/RangeSegment.hpp [hunk 2]
```diff
@@ -267,18 +267,18 @@ struct TypedRangeStrideSegment {
...
-      : m_begin(iterator(begin, stride)),
-        m_end(iterator(end, stride)),
+  RAJA_HOST_DEVICE TypedRangeStrideSegment(Index_type begin, Index_type end, Index_type stride)
+      : m_begin(iterator(DiffT{begin}, DiffT{stride})),
+        m_end(iterator(DiffT{end}, DiffT{stride})),
         // essentially a ceil((end-begin)/stride) but using integer math,
         // and allowing for negative strides
-        m_size((end - begin + stride - (stride > 0 ? 1 : -1)) / stride)
+        m_size((value_type{end} - value_type{begin} + value_type{stride}
+                - (stride > 0 ? value_type{1} : value_type{-1}))
```
Comment (rhornung67) [inline include/RAJA/index/RangeSegment.hpp]: do we need 'value_type' for the stride here?

File: test/unit/cpu/test-segments.cpp [hunk 3]
```diff
@@ -87,13 +87,41 @@ TEST(RangeStrideSegmentTest, sizes_primes)
   RAJA::RangeStrideSegment segment1(0, 7, 3);  // should produce 0,3,6
   ASSERT_EQ(segment1.size(), 3);
 
-  RAJA::RangeStrideSegment segment2(0, 13, 3); // shoudl produce 0,3,6,9,12
+  RAJA::RangeStrideSegment segment2(0, 13, 3); // should produce 0,3,6,9,12
```
Comment (rhornung67) [inline test/unit/cpu/test-segments.cpp]: Good catch!

**Comment by davidbeckingsale** [comment]
👍

**Pull request merged**